### ISSUE TITLE
Data Parallel Implementation Improvements

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#21144]
After #20910, data parallel implementation still needs the following improvements:

- [ ] Fix C++ data parallel for BN
- [ ] Make sure C++ data parallel work for double backward
- [ ] Move `reduce_add` and `reduce_add_coalesced` in `torch/cuda/comm.py` to C++.
- [ ] Move `Broadcast` and `ReduceAddCoalesced` from `torch/nn/parallel/_functions.py` to C++.
- [ ] Make C++ data parallel use `ReduceAddCoalesced`.
- [ ] Consolidate C++ and Python module replicate.

**Comment by flybass** [comment]
Hi,

I’ve been using the c++ api and wanted to change my code to use data parallel. Can you please clarify what the issue is for Batchnormalization and whether it is actually still open?

Thank you,
Daniel
